keps/sig-autoscaling/20191111-autoscaling-with-multiple-thresholds.md

```diff
@@ -0,0 +1,202 @@
+
+First and foremost, upon making a measurement (i.e. querying the Metrics Server or Custom/External Metrics Server) we compare the value to both thresholds, we can only suggest upscaling if the value is at least higher than the high threshold and a downscale event if it is at least lower than the low threshold.
+
+As we add a second threshold, it seems confusing and error prone to keep the logic of the `targetAverageValue` and the `targetValue` nomenclature for those two thresholds. e.g. have the LowThreshold/LowAverageThreshold and deal with cases where a user could set a HighThreshold and a LowAverageThreshold. 
+So we decided to implement a separate option in the spec, `algorithm`.
+
+Depending on your use case, you might want to consider one or the other. The naming of the algorithm is from the point of view of the controller.
+
+1. `average`
+    The ratio used to compare to the thresholds is `value from the metrics provider` / `current number of replicas`, it is compared to the bounds and the recommended number of replicas is `value from the metrics provider` / `threshold` (low or high depending on the current value).
+    - The `average` algorithm is a good fit if you use a metric that does not directly depend on the number of replicas. Typically, the sum of requests received by a Load Balancer can indicate how many webservers we want to have given that we know that a single webserver should handle X rq/s.
+    Adding a replica will not increase/decrease the number of requests received.
+
+2. `absolute`
+    The default value is `absolute`, we compare the raw metric retrieved as is from the Metrics Provider to the thresholds and it is used as the utilization value. The recommended number of replicas is computed as `current number of replicas` * `value from the metrics provider` / `threshold`.
+    - The `absolute` algorithm is the default as it represents the most common use case: You want your application to run between 60% and 80% of CPU, if the metric (avg:cpu.usage) retrieved is at 85%, you need to scale up. The metric has to be correlated to the number of replicas.
+
+This yields the same behavior as the HPA controller where `TargetValue` is equivalent to `absolute` and `TargetAverageValue` is the `average` algorithm.
+
+It is worth noting that in the upstream controller only the `math.Ceil` function is used to compute the recommended number of replicas.
+This means that if you have a threshold at 10, you will need to reach a utilization of 8.999... from the Metrics Provider to downscale to 9 replicas but 10.001 will suffice trigger an upscale event to 11 replicas.
+
+In order to guaratee a symetry of scaling, our implementation of this custom controller uses `math.Floor` if the value is under the Lower bound. We are not suggesting porting this to the upstream controller, but we think it is important to bring it up to have the opportunity to discuss if it is and why it would be a desired behavior.
```

Comment:
I would recommend moving this note to the out-of-scope section.

Reply:
That makes sense - Thanks